src/plugins/shape-generator/index.ts

```diff
@@ -0,0 +1,407 @@
+          latex: "x_{rectangleGenerator}",
+        });
+        const yHelper = this.calc.HelperExpression({
+          latex: "y_{rectangleGenerator}",
+        });
+        const wHelper = this.calc.HelperExpression({
+          latex: "w_{rectangleGenerator}",
+        });
+        const hHelper = this.calc.HelperExpression({
+          latex: "h_{rectangleGenerator}",
+        });
+        const angleHelper = this.calc.HelperExpression({
+          latex: "A_{rectangleGenerator}",
+        });
+
+        angleHelper.observe("numericValue", () => {
+          const x = xHelper.numericValue;
+          const y = yHelper.numericValue;
+          const w = wHelper.numericValue;
+          const h = hHelper.numericValue;
+          const angle = angleHelper.numericValue;
+
+          if (isNaN(x) || isNaN(y) || isNaN(w) || isNaN(h) || isNaN(angle)) {
+            return;
+          }
+
+          // TODO: HelperExpression is missing unobserve types
+          angleHelper.unobserve("numericValue");
+
+          // Create new empty expression to add the rectangle to
+          this.calc.controller.dispatch({
+            type: "new-expression",
+          });
+
+          // Generate rectangle LaTeX
+          const latex = rectangleLatex(w, h, x, y, angle);
+
+          // Set the expression
+          this.calc.setExpression({
+            ...expr,
+            id: this.calc.selectedExpressionId,
+            type: "expression",
+            latex,
+          });
+
+          this.cleanupExpressions();
+          this.isEditingShape = false;
+        });
+      },
+    },
+  ];
+
+  private expressionsMutationObserver: MutationObserver | null = null;
+
+  afterEnable() {
+    // Construct stylesheet
+    const style = document.createElement("style");
+    style.textContent = `
+      /* Disable the save button while editing a shape */
+      [data-shape-generator-is-editing-shape="true"] .save-btn-container .dcg-action-save {
+        opacity: 0.5;
+        pointer-events: none;
+      }
+
+      /* Disable pointer events on disabled Add expression dropdown items */
+      .dcg-add-expression-dropdown div.dcg-popover-interior .dcg-new-item.dcg-disabled {
+        pointer-events: none;
+      }
+
+      /* Hide shape points (controls) and helper functions expressions */
+      .dcg-expressionitem[expr-id^="shape-generator-"][expr-id$="-point"],
+      .dcg-expressionitem[expr-id^="shape-generator-"][expr-id*="-helper"] {
+        display: none !important;
+      }
+
+      /* Hide sliders if the setting is disabled */
+      [data-shape-generator-show-sliders="false"] .dcg-expressionitem[expr-id^="shape-generator-"]:has(.dcg-slider) {
+        display: none !important;
+      }
+
+      .shape-generator-ok-btn {
+        margin-top: 36px;
+        margin-left: 1px;
+      }
+    `;
+    style.id = "shape-generator-styles";
+    document.head.appendChild(style);
+
+    this._onDispatchHandler = this.calc.controller.dispatcher.register(
```

Comment:
Ah sorry I should've pinged you about this. In https://github.com/DesModder/DesModder/pull/909/files I added a way to make custom dispatch handling easier (hopefully) to work with. Using that API would in particular make the undo/redo behavior correct by default.